sklearn/ensemble/_hist_gradient_boosting/_predictor.pyx

```diff
@@ -17,44 +17,59 @@ from .common cimport Y_DTYPE_C
 from .common import Y_DTYPE
 from .common cimport X_BINNED_DTYPE_C
 from .common cimport node_struct
+from ._bitset cimport in_bitset
 
 
 def _predict_from_numeric_data(
         node_struct [:] nodes,
         const X_DTYPE_C [:, :] numeric_data,
+        const X_BINNED_DTYPE_C [:, :] categorical_data,
+        const long[:] orig_feature_to_binned_cat,
         Y_DTYPE_C [:] out):
 
     cdef:
         int i
 
     for i in prange(numeric_data.shape[0], schedule='static', nogil=True):
-        out[i] = _predict_one_from_numeric_data(nodes, numeric_data, i)
+        out[i] = _predict_one_from_numeric_data(
+            nodes, numeric_data, categorical_data,
+            orig_feature_to_binned_cat, i)
 
 
 cdef inline Y_DTYPE_C _predict_one_from_numeric_data(
         node_struct [:] nodes,
         const X_DTYPE_C [:, :] numeric_data,
+        const X_BINNED_DTYPE_C [:, :] categorical_data,
+        const long[:] orig_feature_to_binned_cat,
         const int row) nogil:
     # Need to pass the whole array and the row index, else prange won't work.
     # See issue Cython #2798
 
     cdef:
         node_struct node = nodes[0]
+        long cat_idx
 
     while True:
         if node.is_leaf:
             return node.value
 
-        if isnan(numeric_data[row, node.feature_idx]):
-            if node.missing_go_to_left:
+        if node.is_categorical:
+            cat_idx = orig_feature_to_binned_cat[node.feature_idx]
+            if in_bitset(categorical_data[row, cat_idx], node.cat_threshold):
```

Comment:
does lightgbm also do that? I.e. bin during predict, and rely on a bitset of internally encoded features?

Reply:
lightgbm does not bin during predict. It has a dynamically sized bitset that encodes the input categorical features, so it can accept a category with any cardinality.

Currently, the implementation accepts a category with any cardinality. If the cardinality is higher than `max_bins`, then only the `max_bins` categories are ranked by cardinality and the rest are considered missing. In this way, it is also handling infrequent categories as well. This option is more flexible, but means I have to bin predict which is disappointing.

A simpler alternative would be to restrict the input to be "ints" with range ~ `[0, max_bin]` and anything outside of that range will be considered missing. This would not do anything special to handle infrequent categories, but it will simplify some of the code.

Reply:
I can go either way on this. I spoke to @amueller about this and seems to prefer the current approach of "binning categories during predict".